app/models/custom_field/calculated_value.rb

```diff
@@ -73,29 +67,94 @@ def formula_string
       formula ? formula.fetch("formula", "") : ""
     end
 
+    def formula_referenced_custom_field_ids
+      formula ? formula.fetch("referenced_custom_fields", []) : []
+    end
+
+    def usable_custom_field_references_for_formula
+      visible_cfs = ProjectCustomField
+                      .where(field_format: FIELD_FORMATS_FOR_FORMULA)
+                      .where.not(id:)
+                      .visible
+
+      cache = {}
+      visible_cfs.reject do |custom_field|
+        custom_field.formula_references_id?(id, cache)
+      end
+    end
+
+    def validate_referenced_custom_fields
+      # We can only validate used custom fields from a high-level perspective, since at this point in validation,
+      # we do not have a project context to check against. So we cannot check if the custom fields are actually
+      # enabled for a project and visible to a non-admin user.
+      formula_cfs = formula_referenced_custom_field_ids
+      allowed_cfs = usable_custom_field_references_for_formula.pluck(:id)
+
+      surplus_cfs = formula_cfs - allowed_cfs
+
+      if surplus_cfs.any?
+        custom_field_names = CustomField.where(id: surplus_cfs).pluck(:name)
+        errors.add(:formula, :not_allowed_custom_fields_referenced, custom_fields: custom_field_names.join(", "))
+      end
+    end
+
+    def formula_references_id?(original_id, cache = {})
+      cache.fetch(id) do
+        if field_format_calculated_value?
```

Comment:
`fetch` doesn't assign value, so storing explicitly is needed
```suggestion
        cache[id] = if field_format_calculated_value?
```

Reply:
Sorry about that. I removed that line during testing something and failed to put it back. Fixed.